pkg/network/usm/utils/file_registry_test.go

```diff
@@ -0,0 +1,143 @@
+
+	// Assert that the callback should executed only *once*
+	assert.Equal(t, 1, registerRecorder.CallsForPathID((pathID)))
+
+	// Assert that the two PIDs are being tracked
+	assert.Contains(t, r.GetRegisteredProcesses(), pid1)
+	assert.Contains(t, r.GetRegisteredProcesses(), pid2)
+
+	// Assert that the first call to `Unregister` (from PID 10) doesn't trigger
+	// the callback but removes the PID 10 from the list
+	r.Unregister(pid1)
+	assert.Equal(t, 0, unregisterRecorder.CallsForPathID((pathID)))
+	assert.NotContains(t, r.GetRegisteredProcesses(), pid1)
+	assert.Contains(t, r.GetRegisteredProcesses(), pid2)
+
+	// After the second call to Unregister` we should trigger the callback
+	// because there are no longer PIDs pointing to this file
+	r.Unregister(pid2)
+	assert.Equal(t, 1, unregisterRecorder.CallsForPathID((pathID)))
+	assert.NotContains(t, r.GetRegisteredProcesses(), pid1)
+	assert.NotContains(t, r.GetRegisteredProcesses(), pid2)
+
+	// Telemetry assertions
+	assert.Equal(t, int64(1), r.telemetry.fileRegistered.Get())
+	assert.Equal(t, int64(1), r.telemetry.fileAlreadyRegistered.Get())
+	assert.Equal(t, int64(1), r.telemetry.fileUnregistered.Get())
+}
+
+func TestRepeatedRegistrationsFromSamePID(t *testing.T) {
+	registerRecorder := new(CallbackRecorder)
+	registerCallback := registerRecorder.Fn()
+
+	unregisterRecorder := new(CallbackRecorder)
+	unregisterCallback := unregisterRecorder.Fn()
+
+	r := newFileRegistry()
+	path, pathID := createTempTestFile(t, "foobar")
+	cmd := testutil.OpenFromAnotherProcess(t, path)
+	pid := uint32(cmd.Process.Pid)
+
+	r.Register(path, pid, registerCallback, unregisterCallback)
+	r.Register(path, pid, registerCallback, unregisterCallback)
+	r.Unregister(pid)
+
+	// Assert that despite multiple calls to `Register` from the same PID we
+	// only need a single call to `Unregister` to trigger the unregister callback
+	assert.Equal(t, 1, registerRecorder.CallsForPathID((pathID)))
+	assert.Equal(t, 1, unregisterRecorder.CallsForPathID((pathID)))
+	assert.NotContains(t, r.GetRegisteredProcesses(), pid)
+}
+
+func TestFilePathInCallbackArgument(t *testing.T) {
+	var capturedPath string
+	callback := func(f FilePath) error {
+		capturedPath = f.HostPath
+		return nil
+	}
+
+	path, _ := createTempTestFile(t, "foobar")
+	cmd := testutil.OpenFromAnotherProcess(t, path)
+	pid := cmd.Process.Pid
+
+	r := newFileRegistry()
+	r.Register(path, uint32(pid), callback, callback)
+
+	// Assert that the callback paths match the pattern <proc_root>/<pid>/root/<path>
+	expectedPath := filepath.Join(r.procRoot, strconv.Itoa(pid), "root", path)
+	assert.Equal(t, expectedPath, capturedPath)
+}
+
+func createTempTestFile(t *testing.T, name string) (string, PathIdentifier) {
+	path := filepath.Join(t.TempDir(), name)
+
+	f, err := os.Create(path)
+	require.NoError(t, err)
+	f.Close()
+
+	pathID, err := NewPathIdentifier(path)
+	require.NoError(t, err)
+
+	return path, pathID
```

Comment:
you're leaking the files
please add
```go
t.Cleanup(func() {
    os.Remove(path)
})
```

Reply:
I refactor the code to fix the CI problem, but I'm relying on `t.TempDir()` instead which has the same effect